File: TheWatch-iOS/Services/NotificationResponseHandler.swift
// ===========================================================================
// NotificationResponseHandler — sends user's Accept/Decline responses to backend.
// ===========================================================================
// Handles outbound responses from notification actions back to the Dashboard API.
// All methods are async and designed to be called from Tasks.
//
// In the current mock implementation, responses are logged.
// Live implementation will POST to the ResponseController endpoints.
// ===========================================================================

import Foundation
import CoreLocation

@Observable
final class NotificationResponseHandler {

    static let shared = NotificationResponseHandler()
    private let locationCoordinator = LocationCoordinator.shared

    private init() {}

    // ── Accept ───────────────────────────────────────────────────────────

    /// Accept an SOS dispatch — "I'm on my way."
    /// Escalates location to emergency mode and sends acknowledgment to backend.
    func sendAccept(
        requestId: String,
        notificationId: String,
        incidentLatitude: Double,
        incidentLongitude: Double
    ) async {
        print("[ResponseHandler] ACCEPT for \(requestId)")

        // Escalate location tracking
        locationCoordinator.escalateToEmergency()

        let location = locationCoordinator.currentLocation
        let distance = calculateDistance(
            from: location,
            toLat: incidentLatitude,
            toLon: incidentLongitude
        )

        // TODO: Replace with actual URLSession POST in live implementation
        // POST /api/response/{requestId}/ack
        print("[ResponseHandler] [MOCK] POST /api/response/\(requestId)/ack — distance=\(String(format: "%.0f", distance))m")
    }

    // ── Decline ──────────────────────────────────────────────────────────

    /// Decline an SOS dispatch — "I can't help right now."
    func sendDecline(requestId: String, notificationId: String) async {
        print("[ResponseHandler] DECLINE for \(requestId)")
        print("[ResponseHandler] [MOCK] Decline recorded")
    }

    // ── I'm OK ───────────────────────────────────────────────────────────

    /// Respond "I'm OK" to a check-in request.
    func sendImOk(requestId: String, notificationId: String) async {
        print("[ResponseHandler] IM_OK for \(requestId)")
        print("[ResponseHandler] [MOCK] I'm OK recorded")
    }

    // ── Need Help ────────────────────────────────────────────────────────

    /// Respond "Need Help" — triggers a secondary SOS.
    func sendNeedHelp(
        requestId: String,
        notificationId: String,
        latitude: Double,
        longitude: Double
    ) async {
        print("[ResponseHandler] NEED HELP for \(requestId) — triggering secondary SOS")

        // Escalate our own location
        locationCoordinator.escalateToEmergency()

        // TODO: POST /api/response/trigger with scope=Neighborhood
        print("[ResponseHandler] [MOCK] Secondary SOS triggered from check-in")
    }

    // ── Call 911 ─────────────────────────────────────────────────────────

    /// Log 911 escalation. Phone dialer is launched by NotificationService.
    func sendCall911(requestId: String, notificationId: String) async {
        print("[ResponseHandler] CALL 911 for \(requestId)")
        print("[ResponseHandler] [MOCK] 911 escalation logged")
    }

    // ── Acknowledge ──────────────────────────────────────────────────────

    /// Acknowledge an evacuation notice.
    func sendAcknowledge(requestId: String, notificationId: String) async {
        print("[ResponseHandler] ACKNOWLEDGE for \(requestId)")
        print("[ResponseHandler] [MOCK] Acknowledgment recorded")
    }

    // ── Device Token Registration ────────────────────────────────────────

    /// Register APNs device token with the backend.
    func registerDeviceToken(_ token: String, platform: String) async {
        print("[ResponseHandler] Registering \(platform) token: \(token.prefix(20))...")
        // TODO: POST /api/response/device/register
        print("[ResponseHandler] [MOCK] Device token registered")
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    /// Haversine distance calculation.
    private func calculateDistance(
        from location: CLLocationCoordinate2D?,
        toLat: Double,
        toLon: Double
    ) -> Double {
        guard let loc = location else { return 0 }

        let r = 6371000.0 // Earth radius meters
        let dLat = (toLat - loc.latitude) * .pi / 180
        let dLon = (toLon - loc.longitude) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2) +
                cos(loc.latitude * .pi / 180) * cos(toLat * .pi / 180) *
                sin(dLon / 2) * sin(dLon / 2)
        let c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return r * c
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/service/NotificationResponseHandler.kt
package com.thewatch.app.service

import android.util.Log
import com.thewatch.app.data.repository.UserRepository
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Handles outbound responses to notifications — the user's yes/no answer
 * sent back to the TheWatch backend API.
 *
 * Injected into NotificationActionReceiver and NotificationService.
 * Also handles FCM device token registration.
 *
 * All methods are suspend functions designed to be called from coroutine scopes.
 * In the mock implementation, responses are logged. The live implementation
 * will POST to the Dashboard API's ResponseController endpoints.
 */
@Singleton
class NotificationResponseHandler @Inject constructor(
    private val userRepository: UserRepository,
    private val locationCoordinator: LocationCoordinator
) {
    companion object {
        private const val TAG = "NotifResponseHandler"
    }

    /**
     * Accept an SOS dispatch — "I'm on my way."
     * - Sends acknowledgment to POST /api/response/{requestId}/ack
     * - Escalates location tracking to emergency mode (1-sec updates)
     * - Starts broadcasting location to the response group via SignalR
     */
    suspend fun sendAccept(
        requestId: String,
        notificationId: String,
        incidentLatitude: Double,
        incidentLongitude: Double
    ) {
        Log.i(TAG, "Sending ACCEPT for $requestId")

        // Escalate location to emergency mode so our position is tracked
        locationCoordinator.escalateToEmergency()

        val user = userRepository.getCurrentUser()
        val location = locationCoordinator.getLastKnownLocation()

        // Calculate distance to incident
        val distanceMeters = if (location != null) {
            calculateDistance(
                location.latitude, location.longitude,
                incidentLatitude, incidentLongitude
            )
        } else 0.0

        // TODO: Replace with actual Retrofit API call in live implementation
        // POST /api/response/{requestId}/ack
        // Body: { responderId, responderName, latitude, longitude, distanceMeters, estimatedArrivalMinutes }
        Log.w(TAG, "[MOCK] POST /api/response/$requestId/ack — " +
            "responderId=${user?.id}, distance=${distanceMeters}m")
    }

    /**
     * Decline an SOS dispatch — "I can't help right now."
     */
    suspend fun sendDecline(requestId: String, notificationId: String) {
        Log.i(TAG, "Sending DECLINE for $requestId")

        // TODO: POST to backend to record the decline
        Log.w(TAG, "[MOCK] Decline recorded for $requestId")
    }

    /**
     * Respond "I'm OK" to a check-in request.
     */
    suspend fun sendImOk(requestId: String, notificationId: String) {
        Log.i(TAG, "Sending IM_OK for $requestId")

        // TODO: POST to backend
        Log.w(TAG, "[MOCK] I'm OK recorded for $requestId")
    }

    /**
     * Respond "Need Help" — this is a secondary SOS trigger.
     * The person receiving a check-in actually needs help themselves.
     */
    suspend fun sendNeedHelp(
        requestId: String,
        notificationId: String,
        latitude: Double,
        longitude: Double
    ) {
        Log.w(TAG, "NEED HELP received for $requestId — triggering secondary SOS")

        // Escalate our own location
        locationCoordinator.escalateToEmergency()

        // TODO: POST /api/response/trigger with scope=Neighborhood
        // This creates a NEW response request for the person who needs help
        Log.w(TAG, "[MOCK] Secondary SOS triggered from check-in response")
    }

    /**
     * Call 911 action — logs the escalation. Dialer launch is handled by the receiver.
     */
    suspend fun sendCall911(requestId: String, notificationId: String) {
        Log.w(TAG, "CALL 911 for $requestId — escalation logged")

        // TODO: POST to backend to record 911 escalation event
        Log.w(TAG, "[MOCK] 911 escalation recorded for $requestId")
    }

    /**
     * Acknowledge an evacuation notice — "I see it, I'm evacuating."
     */
    suspend fun sendAcknowledge(requestId: String, notificationId: String) {
        Log.i(TAG, "Sending ACKNOWLEDGE for $requestId")

        // TODO: POST to backend
        Log.w(TAG, "[MOCK] Acknowledgment recorded for $requestId")
    }

    /**
     * Register FCM device token with the backend.
     * Called when FirebaseMessagingService.onNewToken() fires.
     */
    suspend fun registerDeviceToken(token: String) {
        val user = userRepository.getCurrentUser()
        Log.i(TAG, "Registering FCM token for user ${user?.id}: ${token.take(20)}...")

        // TODO: POST /api/response/device/register
        // Body: { userId, deviceToken, platform: "ANDROID", deviceName }
        Log.w(TAG, "[MOCK] Device token registered")
    }

    /**
     * Haversine distance calculation (meters).
     */
    private fun calculateDistance(
        lat1: Double, lon1: Double,
        lat2: Double, lon2: Double
    ): Double {
        val r = 6371000.0 // Earth radius in meters
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                Math.sin(dLon / 2) * Math.sin(dLon / 2)
        val c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
        return r * c
    }
}
